R/helpers/template_utils.R

```diff
@@ -159,3 +167,159 @@ extract_pdf_text <- function(pdf_path) {
+    if (toc[[i]]$level <= start_level) {
+      return(toc[[i]]$title)
+    }
+  }
+
+  return(NULL)
+}
+
+# Safe null-coalescing operator
+`%||%` <- function(a, b) if (!is.null(a)) a else b
+
+#' Normalize Text for Consistent Matching
+#'
+#' @param x A character vector or string to normalize.
+#'
+#' @return A character vector with normalized text.
+normalize_text <- function(x) {
+  x <- gsub("[\r\n\t]", " ", x)
+  x <- gsub("[‘’´`]", "'", x)
+  x <- gsub("[“”]", "\"", x)
+  x <- gsub("–|—", "-", x)
+  x <- gsub("\\s+", " ", x)
+  tolower(trimws(x))
+}
+
+#' Extract question content using PDF bookmarks (outline headings) using PyMuPDF
+#'
+#' @param pdf_path Path to the PDF file
+#' @param heading Title of the heading to extract (e.g., "Question 1")
+#' @return Text block under that heading, up to the next heading
+extract_question_from_pdf <- function(pdf_path, heading) {
+
+  if (!file.exists(pdf_path)) {
+    stop("File not found:", pdf_path)
+  }
+
+  # Load and flatten TOC
+  toc_full <- pdf_toc(pdf_path)
+  toc <- flatten_toc(toc_full$children)
+
+  norm_titles <- normalize_text(sapply(toc, `[[`, "title"))
+  norm_heading <- normalize_text(heading)
+
+  matches <- which(norm_titles == norm_heading)
+  if (length(matches) == 0) {
+    stop("Heading '", heading, "' not found in TOC")
+  }
+
+  next_heading_title <- get_next_heading_title(toc, heading)
+  # Read full text
+  full_text <- pdf_text(pdf_path)
+  lines <- unlist(strsplit(paste(full_text, collapse = "\n"), "\n"))
+  norm_lines <- normalize_text(trimws(lines))
+
+  # Locate start of heading
+  start_idx <- which(norm_lines == norm_heading)
+  if (length(start_idx) == 0) {
+    stop("Heading '", heading, "' not found in text")
+  }
+
+  start_line <- start_idx[1]
+  end_line <- length(lines)
+
+  # Locate next heading line
+  if (!is.null(next_heading_title)) {
+    norm_next <- normalize_text(next_heading_title)
+    next_idx <- which(norm_lines == norm_next)
+    if (length(next_idx) > 0 && next_idx[1] > start_line) {
+      end_line <- next_idx[1] - 1
+    }
+  }
+
+  # Defensive bounds check
+  if (start_line > end_line || start_line < 1 || end_line < 1) {
+    stop("Invalid line bounds: start =", start_line, ", end =", end_line)
+  }
+
+  extracted <- lines[start_line:end_line]
+  return(paste(trimws(extracted), collapse = "\n"))
+}
+
+
+#' Extract a specific question block from a submission file
+#' @param submission Path to the submission file
+#' @param question The exact heading string to look for
+#' @return Text block belonging to the specified question
+extract_question_from_txt <- function(submission, question) {
+  if (!file.exists(submission)) {
+    return(paste0("[Error: Submission file ", basename(submission), " not found]"))
+  }
+
+  lines <- readLines(submission, warn = FALSE)
+
+  # Find the exact heading that matches the question identifier
+  start_idx <- grep(paste0("^\\s*", question, "\\b"), lines, ignore.case = TRUE)
```

Comment:
I'm a bit puzzled by this regex. I don't see how it'll match Markdown headers (`#`, `##, etc.), but I'd like to make sure we're only checking those.